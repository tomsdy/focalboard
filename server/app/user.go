package app

import "github.com/mattermost/focalboard/server/model"

func (a *App) GetTeamUsers(teamID string) ([]*model.User, error) {
	return a.store.GetUsersByTeam(teamID)
}

func (a *App) SearchTeamUsers(teamID string, searchQuery string) ([]*model.User, error) {
	return a.store.SearchUsersByTeam(teamID, searchQuery)
}

func (a *App) UpdateUserConfig(userID string, patch model.UserPropPatch) (map[string]interface{}, error) {
	if err := a.store.PatchUserProps(userID, patch); err != nil {
		return nil, err
	}

	user, err := a.store.GetUserByID(userID)
	if err != nil {
		return nil, err
	}

	return user.Props, nil
}
