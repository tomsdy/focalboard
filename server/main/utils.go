package main

import (
	"crypto/rand"
	"fmt"
	"log"
	"os"
)

// FileExists returns true if a file exists at the path
func fileExists(path string) bool {
	_, err := os.Stat(path)
	if os.IsNotExist(err) {
		return false
	}
	return err == nil
}

// CreateGUID returns a random GUID
func createGUID() string {
	b := make([]byte, 16)
	_, err := rand.Read(b)
	if err != nil {
		log.Fatal(err)
	}
	uuid := fmt.Sprintf("%x-%x-%x-%x-%x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:])

	return uuid
}
