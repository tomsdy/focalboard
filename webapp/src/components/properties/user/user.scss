.UserProperty {
    margin-right: 20px;
    min-width: 120px;

    &.readonly {
        overflow: hidden;
        text-overflow: ellipsis;
        min-width: unset;
    }
}
