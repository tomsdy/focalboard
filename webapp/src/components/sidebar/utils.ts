// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {CategoryBoards, DefaultCategory} from '../../store/sidebar'

import {Block} from '../../blocks/block'
import {Board} from '../../blocks/board'

export function addMissingItems(sidebarCategories: Array<CategoryBoards>, allItems: Array<Block | Board>): Array<CategoryBoards> {
    const blocksInCategories = new Map<string, boolean>()
    sidebarCategories.forEach(
        (category) => category.boardIDs.forEach(
            (boardID) => blocksInCategories.set(boardID, true),
        ),
    )

    const defaultCategory: CategoryBoards = {
        ...DefaultCategory,
        boardIDs: [],
    }

    allItems.forEach((block) => {
        if (!blocksInCategories.get(block.id)) {
            defaultCategory.boardIDs.push(block.id)
        }
    })

    // sidebarCategories comes from store,
    // so is frozen and can't be extended.
    // So creating new array from it.
    return [
        ...sidebarCategories,
        defaultCategory,
    ]
}
