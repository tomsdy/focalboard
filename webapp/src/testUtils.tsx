// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'

export const wrapIntl = (children: any) => <IntlProvider locale='en'>{children}</IntlProvider>

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()

        range.getBoundingClientRect = jest.fn()

        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }

        return range
    }
}
