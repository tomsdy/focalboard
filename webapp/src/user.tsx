// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'

const UserContext = React.createContext(undefined as IUser|undefined)

const WorkspaceUsersContext = React.createContext({
    users: new Array<IUser>(),
    usersById: new Map<string, IUser>(),
})

interface IUser {
    id: string,
    username: string,
    email: string,
    props: Record<string, any>,
    createAt: number,
    updateAt: number,
}

type WorkspaceUsersContextData = {
    users: Array<IUser>
    usersById: Map<string, IUser>
}

export {IUser, UserContext, WorkspaceUsersContext, WorkspaceUsersContextData}
