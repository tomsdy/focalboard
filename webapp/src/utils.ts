// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import marked from 'marked'

declare global {
    interface Window {
        msCrypto: Crypto
    }
}

class Utils {
    static createGuid(): string {
        const crypto = window.crypto || window.msCrypto
        function randomDigit() {
            if (crypto && crypto.getRandomValues) {
                const rands = new Uint8Array(1)
                crypto.getRandomValues(rands)
                return (rands[0] % 16).toString(16)
            }

            return (Math.floor((Math.random() * 16))).toString(16)
        }
        return 'xxxxxxxx-xxxx-4xxx-8xxx-xxxxxxxxxxxx'.replace(/x/g, randomDigit)
    }

    static htmlToElement(html: string): HTMLElement {
        const template = document.createElement('template')
        template.innerHTML = html.trim()
        return template.content.firstChild as HTMLElement
    }

    static getElementById(elementId: string): HTMLElement {
        const element = document.getElementById(elementId)
        Utils.assert(element, `getElementById "${elementId}$`)
        return element!
    }

    static htmlEncode(text: string): string {
        return String(text).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;')
    }

    static htmlDecode(text: string): string {
        return String(text).replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>').replace(/&quot;/g, '"')
    }

    // Markdown

    static htmlFromMarkdown(text: string): string {
        // HACKHACK: Somehow, marked doesn't encode angle brackets
        const renderer = new marked.Renderer()
        renderer.link = (href, title, contents) => `<a target="_blank" rel="noreferrer" href="${href}" title="${title || ''}" onclick="event.stopPropagation();">${contents}</a>`
        const html = marked(text.replace(/</g, '&lt;'), {renderer, breaks: true})
        return html
    }

    // Date and Time

    static displayDate(date: Date): string {
        const dateTimeFormat = new Intl.DateTimeFormat('en', {year: 'numeric', month: 'short', day: '2-digit'})
        const text = dateTimeFormat.format(date)

        return text
    }

    static displayDateTime(date: Date): string {
        const dateTimeFormat = new Intl.DateTimeFormat(
            'en',
            {
                year: 'numeric',
                month: 'short',
                day: '2-digit',
                hour: 'numeric',
                minute: 'numeric',
            })
        const text = dateTimeFormat.format(date)
        return text
    }

    static sleep(miliseconds: number): Promise<void> {
        return new Promise((resolve) => setTimeout(resolve, miliseconds))
    }

    // Errors

    static assertValue(valueObject: any): void {
        const name = Object.keys(valueObject)[0]
        const value = valueObject[name]
        if (!value) {
            Utils.logError(`ASSERT VALUE [${name}]`)
        }
    }

    static assert(condition: any, tag = ''): void {
        /// #!if ENV !== "production"
        if (!condition) {
            Utils.logError(`ASSERT [${tag ?? new Error().stack}]`)
        }

        /// #!endif
    }

    static assertFailure(tag = ''): void {
        /// #!if ENV !== "production"
        Utils.assert(false, tag)

        /// #!endif
    }

    static log(message: string): void {
        /// #!if ENV !== "production"
        const timestamp = (Date.now() / 1000).toFixed(2)
        // eslint-disable-next-line no-console
        console.log(`[${timestamp}] ${message}`)

        /// #!endif
    }

    static logError(message: string): void {
        /// #!if ENV !== "production"
        const timestamp = (Date.now() / 1000).toFixed(2)
        // eslint-disable-next-line no-console
        console.error(`[${timestamp}] ${message}`)

        /// #!endif
    }

    // favicon

    static setFavicon(icon?: string): void {
        const href = icon ? `data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100"><text y=".9em" font-size="90">${icon}</text></svg>` : ''
        const link = (document.querySelector("link[rel*='icon']") || document.createElement('link')) as HTMLLinkElement
        link.type = 'image/x-icon'
        link.rel = 'shortcut icon'
        link.href = href
        document.getElementsByTagName('head')[0].appendChild(link)
    }

    // URL

    static replaceUrlQueryParam(paramName: string, value?: string): void {
        const queryString = new URLSearchParams(window.location.search)
        const currentValue = queryString.get(paramName) || ''
        if (currentValue !== value) {
            const newUrl = new URL(window.location.toString())
            if (value) {
                newUrl.searchParams.set(paramName, value)
            } else {
                newUrl.searchParams.delete(paramName)
            }
            window.history.pushState({}, document.title, newUrl.toString())
        }
    }

    // File names

    static sanitizeFilename(filename: string): string {
        // TODO: Use an industrial-strength sanitizer
        let sanitizedFilename = filename
        const illegalCharacters = ['\\', '/', '?', ':', '<', '>', '*', '|', '"', '.']
        illegalCharacters.forEach((character) => {
            sanitizedFilename = sanitizedFilename.replace(character, '')
        })
        return sanitizedFilename
    }

    // File picker

    static selectLocalFile(onSelect?: (file: File) => void, accept = '.jpg,.jpeg,.png'): void {
        const input = document.createElement('input')
        input.type = 'file'
        input.accept = accept
        input.onchange = async () => {
            const file = input.files![0]
            onSelect?.(file)
        }

        input.style.display = 'none'
        document.body.appendChild(input)
        input.click()

        // TODO: Remove or reuse input
    }

    // Arrays

    static arraysEqual(a: readonly any[], b: readonly any[]): boolean {
        if (a === b) {
            return true
        }
        if (a === null || b === null) {
            return false
        }
        if (a === undefined || b === undefined) {
            return false
        }
        if (a.length !== b.length) {
            return false
        }

        for (let i = 0; i < a.length; ++i) {
            if (a[i] !== b[i]) {
                return false
            }
        }
        return true
    }

    static arrayMove(arr: any[], srcIndex: number, destIndex: number): void {
        arr.splice(destIndex, 0, arr.splice(srcIndex, 1)[0])
    }

    // Clipboard

    static copyTextToClipboard(text: string): boolean {
        const textField = document.createElement('textarea')
        textField.innerText = text
        textField.style.position = 'fixed'
        textField.style.opacity = '0'

        document.body.appendChild(textField)
        textField.select()

        let result = false
        try {
            result = document.execCommand('copy')
        } catch (err) {
            Utils.logError(`copyTextToClipboard ERROR: ${err}`)
            result = false
        }
        textField.remove()

        return result
    }

    static isMobile() {
        const toMatch = [
            /Android/i,
            /webOS/i,
            /iPhone/i,
            /iPad/i,
            /iPod/i,
            /BlackBerry/i,
            /Windows Phone/i,
        ]

        return toMatch.some((toMatchItem) => {
            return navigator.userAgent.match(toMatchItem)
        })
    }
}

export {Utils}
