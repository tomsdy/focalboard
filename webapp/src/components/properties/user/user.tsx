// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useContext} from 'react'
import Select from 'react-select'

import {IUser, WorkspaceUsersContext, WorkspaceUsers} from '../../../user'

import './user.scss'
import {getSelectBaseStyle} from '../../../theme'

type Props = {
    value: string,
    readonly: boolean,
    onChange: (value: string) => void,
}

const UserProperty = (props: Props): JSX.Element => {
    const workspaceUsers = useContext<WorkspaceUsers>(WorkspaceUsersContext)

    if (props.readonly) {
        return (<div className='UserProperty octo-propertyvalue readonly'>{workspaceUsers.usersById.get(props.value)?.username || props.value}</div>)
    }

    return (
        <Select
            options={workspaceUsers.users}
            isSearchable={true}
            isClearable={true}
            backspaceRemovesValue={true}
            className={'UserProperty'}
            styles={getSelectBaseStyle()}
            getOptionLabel={(o: IUser) => o.username}
            getOptionValue={(a: IUser) => a.id}
            value={workspaceUsers.usersById.get(props.value) || null}
            onChange={(item, action) => {
                if (action.action === 'select-option') {
                    props.onChange(item?.id || '')
                } else if (action.action === 'clear') {
                    props.onChange('')
                }
            }}
        />
    )
}

export default UserProperty
