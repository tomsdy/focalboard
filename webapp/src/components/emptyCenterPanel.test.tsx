// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {render, screen, waitFor} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import {MockStoreEnhanced} from 'redux-mock-store'

import {mocked} from 'ts-jest/utils'

import {Provider as ReduxProvider} from 'react-redux'

import {MemoryRouter} from 'react-router-dom'

import Mutator from '../mutator'

import {Utils} from '../utils'

import {UserWorkspace} from '../user'

import {mockDOM, mockStateStore, wrapDNDIntl} from '../testUtils'

import EmptyCenterPanel from './emptyCenterPanel'

jest.mock('react-router-dom', () => {
    const originalModule = jest.requireActual('react-router-dom')

    return {
        ...originalModule,
        useRouteMatch: jest.fn(() => {
            return {url: '/'}
        }),
    }
})
jest.mock('../utils')
jest.mock('../mutator')

describe('components/emptyCenterPanel', () => {
    const mockedUtils = mocked(Utils, true)
    const mockedMutator = mocked(Mutator, true)
    const workspace1: UserWorkspace = {
        id: 'workspace_1',
        title: 'Workspace 1',
        boardCount: 1,
    }
    const template1Title = 'Template 1'
    const globalTemplateTitle = 'Template Global'
    let store:MockStoreEnhanced<unknown, unknown>
    beforeAll(mockDOM)
    beforeEach(() => {
        jest.clearAllMocks()
        const state = {
            workspace: {
                userWorkspaces: new Array<UserWorkspace>(workspace1),
                current: workspace1,
            },
            boards: {
                templates: [
                    {id: '1', title: template1Title, fields: {icon: '🚴🏻‍♂️'}},
                ],
            },
            globalTemplates: {
                value: [{id: 'global-1', title: globalTemplateTitle, fields: {icon: '🚴🏻‍♂️'}}],
            },
        }
        store = mockStateStore([], state)
    })
    describe('not a focalboard Plugin', () => {
        beforeAll(() => {
            mockedUtils.isFocalboardPlugin.mockReturnValue(false)
        })
        test('should match snapshot', () => {
            const {container} = render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <EmptyCenterPanel/>
                </ReduxProvider>
                ,
            ), {wrapper: MemoryRouter})
            expect(container).toMatchSnapshot()
        })
    })
    describe('a focalboard Plugin', () => {
        beforeAll(() => {
            mockedUtils.isFocalboardPlugin.mockReturnValue(true)
        })
        test('should match snapshot', () => {
            const {container} = render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <EmptyCenterPanel/>
                </ReduxProvider>
                ,
            ), {wrapper: MemoryRouter})
            expect(container).toMatchSnapshot()
        })
        test('return emptyCenterPanel and click new template', () => {
            const {container} = render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <EmptyCenterPanel/>
                </ReduxProvider>
                ,
            ), {wrapper: MemoryRouter})
            const divNewTemplate = container.querySelector('div.button.new-template')
            expect(divNewTemplate).not.toBeNull()
            userEvent.click(divNewTemplate!)
            expect(mockedMutator.insertBlocks).toBeCalledTimes(1)
        })
        test('return emptyCenterPanel and click empty board', () => {
            render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <EmptyCenterPanel/>
                </ReduxProvider>
                ,
            ), {wrapper: MemoryRouter})
            const divEmptyboard = screen.getByText('Start with an Empty Board').parentElement
            expect(divEmptyboard).not.toBeNull()
            userEvent.click(divEmptyboard!)
            expect(mockedMutator.insertBlocks).toBeCalledTimes(1)
        })
        test('return emptyCenterPanel and click to add board from template', async () => {
            render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <EmptyCenterPanel/>
                </ReduxProvider>
                ,
            ), {wrapper: MemoryRouter})
            const divAddBoard = screen.getByText(template1Title).parentElement
            expect(divAddBoard).not.toBeNull()
            userEvent.click(divAddBoard!)
            await waitFor(async () => {
                expect(mockedMutator.duplicateBoard).toBeCalledTimes(1)
            })
        })
        test('return emptyCenterPanel and click to add board from global template', async () => {
            render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <EmptyCenterPanel/>
                </ReduxProvider>
                ,
            ), {wrapper: MemoryRouter})
            const divAddBoard = screen.getByText(globalTemplateTitle).parentElement
            expect(divAddBoard).not.toBeNull()
            userEvent.click(divAddBoard!)
            await waitFor(async () => {
                expect(mockedMutator.duplicateFromRootBoard).toBeCalledTimes(1)
            })
        })
    })
})
