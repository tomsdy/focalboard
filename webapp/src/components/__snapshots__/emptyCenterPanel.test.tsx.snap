// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/emptyCenterPanel a focalboard Plugin should match snapshot 1`] = `
<div>
  <div
    class="EmptyCenterPanel"
  >
    <div
      class="content"
    >
      <span
        class="title"
      >
        Create a Board in Workspace 1
      </span>
      <span
        class="description"
      >
        Add a board to the sidebar using any of the templates defined below or start from scratch.
        <br />
         Members of "
        <b>
          Workspace 1
        </b>
        " will have access to boards created here.
      </span>
      <span
        class="choose-template-text"
      >
        Choose a template
      </span>
      <div
        class="button-container"
      >
        <div
          class="button "
        >
          <span>
            üö¥üèª‚Äç‚ôÇÔ∏è
          </span>
          <span
            class="button-title"
          >
            Template 1
          </span>
          <div
            aria-label="menuwrapper"
            class="MenuWrapper"
            role="button"
          >
            <button
              class="Button IconButton"
              type="button"
            >
              <i
                class="CompassIcon icon-dots-horizontal OptionsIcon"
              />
            </button>
          </div>
        </div>
        <div
          class="button "
        >
          <span>
            üö¥üèª‚Äç‚ôÇÔ∏è
          </span>
          <span
            class="button-title"
          >
            Template Global
          </span>
        </div>
        <div
          class="button new-template"
        >
          <span>
            <i
              class="CompassIcon icon-plus AddIcon"
            />
          </span>
          <span
            class="button-title"
          >
            New template
          </span>
        </div>
      </div>
      <span
        class="choose-template-text"
      >
        or
      </span>
      <div
        class="button "
      >
        <span>
          <svg
            class="BoardIcon Icon"
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="80"
              rx="5"
              ry="5"
              width="80"
              x="10"
              y="10"
            />
            <polyline
              points="28,25 28,55"
              style="stroke-width: 15px;"
            />
            <polyline
              points="50,25 50,70"
              style="stroke-width: 15px;"
            />
            <polyline
              points="72,25 72,45"
              style="stroke-width: 15px;"
            />
          </svg>
        </span>
        <span
          class="button-title"
        >
          Start with an Empty Board
        </span>
      </div>
      You can change the channel using the switcher in the sidebar.
    </div>
  </div>
</div>
`;

exports[`components/emptyCenterPanel not a focalboard Plugin should match snapshot 1`] = `
<div>
  <div
    class="EmptyCenterPanel"
  >
    <div
      class="Hint"
    >
      Add or select a board from the sidebar to get started.
    </div>
  </div>
</div>
`;
